File: Unstop/100DayCode/Day 6/SwapLinkedListNode.cpp
#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

ListNode* swapNodes(ListNode* head, int k) {
    // Count the length of the linked list
    int length = 0;
    ListNode* curr = head;
    while (curr) {
        length++;
        curr = curr->next;
    }

    // Find the kth node from the beginning
    ListNode* kth_from_beginning = head;
    for (int i = 1; i < k; ++i) {
        kth_from_beginning = kth_from_beginning->next;
    }

    // Find the kth node from the end
    ListNode* kth_from_end = head;
    for (int i = 0; i < length - k; ++i) {
        kth_from_end = kth_from_end->next;
    }

    // Swap the values
    swap(kth_from_beginning->val, kth_from_end->val);

    return head;
}

void printLinkedList(ListNode* head) {
    ListNode* curr = head;
    while (curr) {
        cout << curr->val << " ";
        curr = curr->next;
    }
    cout << endl;
}

int main() {
    int n;
    cin >> n;
    vector<int> values(n);
    for (int i = 0; i < n; ++i) {
        cin >> values[i];
    }
    int k;
    cin >> k;

    // Create the linked list
    ListNode* head = new ListNode(values[0]);
    ListNode* curr = head;
    for (int i = 1; i < n; ++i) {
        curr->next = new ListNode(values[i]);
        curr = curr->next;
    }

    // Swap nodes
    head = swapNodes(head, k);

    // Output
    printLinkedList(head);

    // Free memory
    while (head) {
        ListNode* temp = head;
        head = head->next;
        delete temp;
    }

    return 0;
}

File: Unstop/100DayCode/Day 6/SwapLinkedListNode.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def swapNodes(head, k):
    # Count the length of the linked list
    length = 0
    curr = head
    while curr:
        length += 1
        curr = curr.next

    # Find the kth node from the beginning
    kth_from_beginning = head
    for _ in range(k - 1):
        kth_from_beginning = kth_from_beginning.next

    # Find the kth node from the end
    kth_from_end = head
    for _ in range(length - k):
        kth_from_end = kth_from_end.next

    # Swap the values
    kth_from_beginning.val, kth_from_end.val = kth_from_end.val, kth_from_beginning.val

    return head

def printLinkedList(head):
    curr = head
    while curr:
        print(curr.val, end=" ")
        curr = curr.next
    print()

# Input
n = int(input())
values = list(map(int, input().split()))
k = int(input())

# Create the linked list
head = ListNode(values[0])
curr = head
for val in values[1:]:
    curr.next = ListNode(val)
    curr = curr.next

# Swap nodes
head = swapNodes(head, k)

# Output
printLinkedList(head)
